### PR TITLE
feat: add rebalanceShortfall and refactor forceRebalance

### PR DESCRIPTION
**Pull request by ev-d**
## Description

Add rebalanceShortfall public function that contains calculations which erlier used in forceRebalance function. It helps UI to reduce calls and calculations. Also, forceRebalance function was refactored and instead calculations it calls rebalanceShortfall function that returns amount to rebalance

## Problem

Number of RPC queries for getting amount to rebalance

## Solution

Provides a number of ETH for rebalancing vault using one public function. Also, helps UI to reduce calculations

### CODE REVIEW

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 1]
```diff
@@ -156,6 +156,32 @@ contract VaultHub is PausableUntilWithRoles {
         ) >= STETH.getPooledEthBySharesRoundUp(socket.sharesMinted);
     }
 
+    /// @notice provides a gap for rebalancing vault
+    /// @param _vault vault address
+    /// @return amount to rebalance
+    function rebalanceShortfall(address _vault) public view returns (uint256) {
```
Comment (tamtamchik) [inline contracts/0.8.25/vaults/VaultHub.sol:172]: Must have a separate tests set to cover all path of the function (reverts, success)

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 2]
```diff
@@ -156,6 +156,32 @@ contract VaultHub is PausableUntilWithRoles {
...
+        // (mintedStETH - X) * BPS_BASE = (vault.valuation() - X) * maxMintableRatio
+        // mintedStETH * BPS_BASE - X * BPS_BASE = vault.valuation() * maxMintableRatio - X * maxMintableRatio
+        // X * maxMintableRatio - X * BPS_BASE = vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE
+        // X * (maxMintableRatio - BPS_BASE) = vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE
+        // X = (vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE) / (maxMintableRatio - BPS_BASE)
+        // X = (mintedStETH * BPS_BASE - vault.valuation() * maxMintableRatio) / (BPS_BASE - maxMintableRatio)
+        // reserveRatio = BPS_BASE - maxMintableRatio
+        // X = (mintedStETH * BPS_BASE - vault.valuation() * maxMintableRatio) / reserveRatio
+
+        return (mintedStETH * TOTAL_BASIS_POINTS - IStakingVault(_vault).valuation() * maxMintableRatio) / socket.reserveRatioBP;
```
Comment (tamtamchik) [inline contracts/0.8.25/vaults/VaultHub.sol]: Maybe cache `socket.reserveRatioBP`?

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 3]
```diff
@@ -316,27 +342,7 @@ contract VaultHub is PausableUntilWithRoles {
         if (_vault == address(0)) revert ZeroArgument("_vault");
         _requireUnhealthy(_vault);
 
-        VaultSocket storage socket = _connectedSocket(_vault);
-
-        uint256 mintedStETH = STETH.getPooledEthByShares(socket.sharesMinted); // TODO: fix rounding issue
-        uint256 reserveRatioBP = socket.reserveRatioBP;
```
Comment (tamtamchik) [inline contracts/0.8.25/vaults/VaultHub.sol:337]: Here it was cached.

File: test/0.8.25/vaults/vaulthub/vaulthub.hub.test.ts [hunk 4]
```diff
@@ -388,6 +388,34 @@ describe("VaultHub.sol:hub", () => {
       expect(await vaultHub.isVaultHealthy(vaultAddress)).to.equal(true);
     });
 
+    it("returns correct value for rebalance vault", async () => {
```
Comment (tamtamchik) [inline test/0.8.25/vaults/vaulthub/vaulthub.hub.test.ts]: Move to special tests for `rebalanceShortfall` context please

**Review by tamtamchik — COMMENTED** [review]
Looks good!

Please add more tests that will cover all the paths within `rebalanceShortfall` function and some tests that will check calculations around boundary values.

**Review by tamtamchik — APPROVED** [review]
🚢

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 5]
```diff
@@ -156,6 +156,33 @@ contract VaultHub is PausableUntilWithRoles {
         ) >= STETH.getPooledEthBySharesRoundUp(socket.sharesMinted);
     }
 
+    /// @notice provides a gap for rebalancing vault
```
Comment (TheDZhon) [inline contracts/0.8.25/vaults/VaultHub.sol]:
```suggestion
    /// @notice estimate ether amount to make the vault healthy using rebalance
```

Comment (TheDZhon) [reply contracts/0.8.25/vaults/VaultHub.sol]: what about bad debt? probably `+inf`

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 6]
```diff
@@ -156,6 +156,33 @@ contract VaultHub is PausableUntilWithRoles {
...
+    /// @param _vault vault address
+    /// @return amount to rebalance
+    function rebalanceShortfall(address _vault) public view returns (uint256) {
+        if (_vault == address(0)) revert ZeroArgument("_vault");
+
+        VaultSocket storage socket = _connectedSocket(_vault);
+
+        uint256 mintedStETH = STETH.getPooledEthByShares(socket.sharesMinted); // TODO: fix rounding issue
```
Comment (TheDZhon) [inline contracts/0.8.25/vaults/VaultHub.sol]: probably should be round up

Comment (TheDZhon) [reply contracts/0.8.25/vaults/VaultHub.sol]: good: overestimate
bad: underestimate

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 7]
```diff
@@ -156,6 +156,33 @@ contract VaultHub is PausableUntilWithRoles {
...
+        uint256 reserveRatioBP = socket.reserveRatioBP;
+        uint256 maxMintableRatio = (TOTAL_BASIS_POINTS - reserveRatioBP);
+
+        // how much ETH should be moved out of the vault to rebalance it to minimal reserve ratio
```
Comment (TheDZhon) [inline contracts/0.8.25/vaults/VaultHub.sol]: probably we don't need it

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 8]
```diff
@@ -156,6 +156,33 @@ contract VaultHub is PausableUntilWithRoles {
...
+
+        // (mintedStETH - X) / (vault.valuation() - X) = maxMintableRatio / BPS_BASE
+        // (mintedStETH - X) * BPS_BASE = (vault.valuation() - X) * maxMintableRatio
```
Comment (TheDZhon) [inline contracts/0.8.25/vaults/VaultHub.sol]: BPS_BASE = TOTAL_BASIS_POINTS

File: contracts/0.8.25/vaults/VaultHub.sol (hunk 5 above)
Comment (TheDZhon) [reply contracts/0.8.25/vaults/VaultHub.sol]: let's handle this case

File: contracts/0.8.25/vaults/VaultHub.sol [hunk 9]
```diff
@@ -156,6 +156,33 @@ contract VaultHub is PausableUntilWithRoles {
...
+        // mintedStETH * BPS_BASE - X * BPS_BASE = vault.valuation() * maxMintableRatio - X * maxMintableRatio
+        // X * maxMintableRatio - X * BPS_BASE = vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE
+        // X * (maxMintableRatio - BPS_BASE) = vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE
+        // X = (vault.valuation() * maxMintableRatio - mintedStETH * BPS_BASE) / (maxMintableRatio - BPS_BASE)
+        // X = (mintedStETH * BPS_BASE - vault.valuation() * maxMintableRatio) / (BPS_BASE - maxMintableRatio)
+        // reserveRatio = BPS_BASE - maxMintableRatio
+        // X = (mintedStETH * BPS_BASE - vault.valuation() * maxMintableRatio) / reserveRatio
+
```
Comment (TheDZhon) [inline contracts/0.8.25/vaults/VaultHub.sol:197]: here we can handle bad debt to not revert below (if condition to return early)

**Review by TheDZhon — CHANGES_REQUESTED** [review]
:clap: Thank you, a bit more polishing would be needed to push it over the line!

File: contracts/0.8.25/vaults/VaultHub.sol (hunk 9 above)
Comment (tamtamchik) [reply contracts/0.8.25/vaults/VaultHub.sol:197]: @TheDZhon can you provide more details on that? I do not see any reverts below

**Review by folkyatina — APPROVED** [review]
LGTM
